src/libs/ck-libs/io/ckio.C

```diff
@@ -398,6 +530,57 @@ namespace Ck { namespace IO {
           return 0;
         }
       };
+
+
+    // class that is used to aggregate the data for a specific read call made by the user
+    class ReadAssembler : public CBase_ReadAssembler {
+	private:
+		std::vector<char> _data_buffer; // the data buffer that is used to store the read request
+		Session _session;
+		size_t _bytes_left; // the number of bytes the user wants for a particular read
+		size_t _read_offset; // the offset they specify for their read
+		CkCallback _after_read; // the callback to invoke after the read is complete
+	public:
+		ReadAssembler(Session session, size_t bytes, size_t offset, CkCallback after_read){
+			_session = session;
+			_bytes_left = bytes;
+			_read_offset = offset;
+			_after_read = after_read;
+			_data_buffer.resize(_bytes_left, 'r'); // resize the buffer to the size of read call
+			_data_buffer.shrink_to_fit();
+			// ckout << "RA Buffer init\n"; 
+			/*
+			for(char ch : _data_buffer)
+				ckout << ch;
+			ckout << endl;
+			std::cout << "ReadAssembler built\n";
+			*/
+		}
+
+		void shareData(size_t read_chare_offset, std::vector<char> data){
+			int start_idx = read_chare_offset - _read_offset; // start index for writing to _data_buffer
+			int counter = 0;
+			for(char& ch : data){ // copy over the data
+			// 	ckout << "ReadAssembler " << read_chare_offset << " has received " << ch << endl;
+				_data_buffer[start_idx + counter] = ch;
+		//		ckout << "Read Assembler " << read_chare_offset << "has set " << (start_idx + counter) << " to " << _data_buffer[start_idx + counter] << endl;
+				++counter;	
+			}
+			_bytes_left -= data.size(); // decrement the number of remaining bytes to read
+			if(_bytes_left) return; // if there are bytes still to read, just return
+			char* buffer = data.data(); 
+			// ckout << "Offset " << _read_offset << " belongs to Pe " << CkMyPe() << endl;
+			ReadCompleteMsg* msg = new (_data_buffer.size()) ReadCompleteMsg();
+			memcpy(msg -> data, buffer, _data_buffer.size());
```

Comment:
This doesn't look right, it's copying from `buffer`, but using `_data_buffer`'s size. Has this been tested with a read that requires data from multiple readers?